design-proposals/volume-claim-templates.md

```diff
@@ -0,0 +1,117 @@
+
+## Goals
+- Provide a method for creating VMs with PVCs that suits best the user's storage while using CDI volume populators.
+
+## Non Goals
+
+## User Stories
+* As a user creating VMs, I want to be able to create a PVC within the VM spec while getting the best PVC spec that is also compatible with backup and restore solutions.
+
+## Repos
+* **CDI**: Provide API to that receives a PVC spec and in case of missing properties such as VolumeMode and AccessModes fills the optimal values according to the storage class.
+* **CDI**: Update DataImportCron API to create/update VolumeCloneSources that can be consumed as sources for PVC handled by CDI volume populators. VolumeCloneSource will need to be able to be updated with different PVC source.
+* **Kubevirt**: Provide API to create a VM with minimal volume information and receive a VM with best suited populated PVC.
+* **Kubevirt**: Provide virtctl command to generate and create PVCs.
+* **SSP**: Manage setting up VolumeCloneSources for commonly supported operating systems updated by DataImportCrons.
+* **Common Templates**: Transition to use VolumeClaimTemplates in VM specs and use CDI's VolumeCloneSource as the source in the template.
+* **OCP UI**: Provide API to create a CDI Volume populator CR that can be used by the volumeClaim. In case of golden image source need to ensure VolumeCloneSource is linked properly to a PVC in order to determine whether the boot image for a common template is available.
+
+# Design
+
+A new field will be added to the VM object called **VolumeClaimTemplates**. This field will provide the user a way to define and create a PVC owned by the VM while providing partial/full PVC spec embedded in the VM spec. In case of partial spec we will make sure the preferred VolumeMode and AccessModes combination according to the StorageProfile associated with provided/default StorageClass are used.  
+This field should replace the DataVolumeTemplates field. Instead of providing a template for a DataVolume which eventually creates one, we will provide a PVC template and create it. DataVolumes and DataVolumeTemplates should be deprecated eventually.  
+A new API will be created in the CDI API library which will receive a partial PVC template and using existing CDI logic will return a complete PVC spec.  
+The provided PVC template in the VolumeClaimTemplate should have a source that can be handled by the new existing CDI volume populators API (DataSourceRef with CDI CR source) or a standalone k8s way (user costume populator, empty disk etc..)
+
+A virtctl command called `create-pvc` can be added. Given minimal flags such as name, size and source this command can generate and create a PVC. The command will use the CDI API similarly as will be done in the VM controller.
```

Comment:
when using PVC that is being populated with CDI volume populators there is no need to wait for it to be bound (for example in case of WFFC) and the VM can be started immediately, VM will be ready when the PVC is bound which is only after the population is over. thats one of the benefits of this new API.